flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java

```diff
@@ -20,249 +20,258 @@
-        final int transformationId = Transformation.getNewNodeId();
-
-        streamGraph.addOperator(
-                transformationId,
-                slotSharingGroup,
-                sinkTransformation.getCoLocationGroupKey(),
-                operatorFactory,
-                inTypeInfo,
-                outTypInfo,
-                name);
+        private <CommT> void addCommittingTopology(Sink<T> sink, DataStream<T> inputStream) {
+            TwoPhaseCommittingSink<T, CommT> committingSink =
+                    (TwoPhaseCommittingSink<T, CommT>) sink;
+            TypeInformation<CommittableMessage<CommT>> typeInformation =
+                    CommittableMessageTypeInfo.of(committingSink::getCommittableSerializer);
+
+            DataStream<CommittableMessage<CommT>> written =
+                    adjustTransformations(
+                            inputStream,
+                            input ->
+                                    input.transform(
+                                            WRITER_NAME,
+                                            typeInformation,
+                                            new SinkWriterOperatorFactory<>(sink)));
+
+            DataStream<CommittableMessage<CommT>> precommitted = addFailOverRegion(written);
+
+            if (sink instanceof WithPreCommitTopology) {
+                precommitted =
+                        adjustTransformations(
+                                precommitted,
+                                ((WithPreCommitTopology<T, CommT>) sink)::addPreCommitTopology);
+            }
+            DataStream<CommittableMessage<CommT>> committed =
+                    adjustTransformations(
+                            precommitted,
+                            pc ->
+                                    pc.transform(
+                                            COMMITTER_NAME,
+                                            typeInformation,
+                                            new CommitterOperatorFactory<>(committingSink)));
+            if (sink instanceof WithPostCommitTopology) {
+                DataStream<CommittableMessage<CommT>> postcommitted = addFailOverRegion(committed);
+                adjustTransformations(
+                        postcommitted,
+                        pc -> {
+                            ((WithPostCommitTopology<T, CommT>) sink).addPostCommitTopology(pc);
+                            return null;
+                        });
+            }
+        }
 
-        streamGraph.setParallelism(transformationId, parallelism);
-        streamGraph.setMaxParallelism(transformationId, maxParallelism);
+        /**
+         * Adds a batch exchange that materializes the output first. This is a no-op in STREAMING.
+         */
+        private <I> DataStream<I> addFailOverRegion(DataStream<I> input) {
+            return new DataStream<>(
+                    executionEnvironment,
+                    new PartitionTransformation<>(
```

Comment:
We might need to consider the cases like pre-commit operators starts with an operator with different parallelism. Perhaps we need to add a new identity map here.

Reply:
Good point, since all custom topologies are marked as experimental. I would take this as a follow-up.

Reply:
It is also acceptable to me~